File: 2017/keeponthegrass.py
# Locate quads
# Track position with quad movement. If touching input coord stop. 
import numpy as np
move_dict = {
    "R" : np.array([-1,0]),
    "L" : np.array([1,0]),
    "D" : np.array([0,1]),
    "U" : np.array([0,-1])
}

def diamond_width(n):
    if n == 0:
        return 1
    if n < 0:
        return 0
    return diamond_width(n-1)*2+1

def locate_quad(coord,centre):
    new_coord = coord[:]
    new_coord[0] -= centre[0]
    new_coord[1] -= centre[1]
    x,y = new_coord
    if y < x and y > -x:
        return "R"
    if y > x and y < -x:
        return "L"
    if y > x and y > -x:
        return "U"
    if y < x and y < -x:
        return "D"
    
def find_quads(coord,centre,n):
    width = round((diamond_width(n)-1)/2/2)
    quad = locate_quad(coord,centre)
    new_centre = centre
    if quad == "R":
        new_centre = [centre[0]+width,centre[1]]
    if quad == "L":
        new_centre = [centre[0]-width,centre[1]]
    if quad == "U":
        new_centre = [centre[0],centre[1]+width]
    if quad == "D":
        new_centre = [centre[0],centre[1]-width]
    return coord,new_centre,quad

File: code/2017/keeponthegrass.cpp
#include <iostream>
#include <math.h> 

using namespace std;

int diamond_width(int n){
    if (n == 0) return 0;
    if (n < 0) return 0;
    return diamond_width(n-1)*2+1;
}

char locate_quad(pair<int,int> coord,pair<int,int> centre){
    pair<int,int> new_coord = coord;
    new_coord.first -= centre.first;
    new_coord.second -= centre.second;
    int x = new_coord.first ;
    int y = new_coord.second;

    if (y < x && y > -x)
        return 'R';
    if (y > x && y < -x)
        return 'L';
    if (y > x && y > -x)
        return 'U';
    if (y < x && y < -x)
        return 'D';
    return 'O';
}

pair<pair<int,int>,char> find_quads(pair<int,int> coord, pair<int,int> centre, int n){
    int width = round(diamond_width(n));
    char quad = locate_quad(coord,centre);
    pair<int,int> new_centre = centre;
    if (quad == 'R')
        new_centre.first = centre.first + width;
    if (quad == 'L')
        new_centre.first = centre.first - width;
    if (quad == 'U')
       new_centre.second = centre.second + width;
    if (quad == 'D')
       new_centre.second = centre.second - width;
    pair<pair<int,int>,char> return_value;
    return_value.first = new_centre;
    return_value.second = quad;
    return return_value;
}

int main(){
    int x,y;
    cin >> x >> y;
    pair<int,int> centre = {0,0};
    pair<int,int> coord;
    coord.first = x;
    coord.second = y;
    int n = 18;
    while (n > 0){
        pair<pair<int,int>,char> return_value = find_quads(coord,centre,n);
    }
    return 0;
}